toolchain/lower/function_context.h

```diff
@@ -18,10 +19,32 @@ namespace Carbon::Lower {
 // `llvm::Function` definition.
 class FunctionContext {
  public:
-  explicit FunctionContext(FileContext& file_context, llvm::Function* function,
-                           SemIR::SpecificId specific_id,
-                           llvm::DISubprogram* di_subprogram,
-                           llvm::raw_ostream* vlog_stream);
+  explicit FunctionContext(
+      FileContext& file_context, llvm::Function* function,
+      SemIR::SpecificId specific_id, llvm::DISubprogram* di_subprogram,
+      llvm::raw_ostream* vlog_stream,
+      FileContext::SpecificFunctionFingerprint* specific_function_state);
+
+  // Describes a function's body fingerprint while creating the function body.
+  // The final fingerprint is stored in the FileContext as a
+  // SpecificFunctionFingerprint.
+  // Create two function fingerprints, where both fingerprints include data
+  // that's evaluated (and hence lowered) differently based on the specific_id.
+  // The first fingerprint includes global values, types and function_id for
+  // functions called inside the function body. The second fingerprint includes
+  // specific_ids for functions called.
+  // For two specifics of the same generic: if the first fingerprint
+  // (`function_common_fingerprint`) is different, the specifics cannot be
+  // coalesced; if the first and second fingerprint
+  // (`function_specific_fingerprint`) are the same, the specifics can be
+  // coalesced without additional checks. If the `function_common_fingerprint`
+  // is the same but `function_specific_fingerprint` is different, additional
+  // checks are needed, i.e. inspecting the non-hashed specific_ids.
+  // TODO: Consider optimizations for repeated entries in both fingerprints.
+  struct LoweringFunctionFingerprint {
+    llvm::BLAKE3 function_common_fingerprint;
```

Comment:
Minor, but you could probably rename these fields to `common_fingerprint`/`specific_fingerprint` or even just `common`/`specific` without losing much readability.